pkg/migration/migrations/retry_jobs_with_exponential_backoff.go

```diff
@@ -0,0 +1,271 @@
+		// Retrieve the table.
+		jt, err := readTableDescriptor(ctx, d, tableID)
+		if err != nil {
+			return err
+		}
+		// Wait for any in-flight schema changes to complete.
+		if mutations := jt.GetMutationJobs(); len(mutations) > 0 {
+			for _, mutation := range mutations {
+				log.Infof(ctx, "waiting for the mutation job %v to complete", mutation.JobID)
+				if d.TestingKnobs.BeforeWaitInRetryJobsWithExponentialBackoffMigration != nil {
+					d.TestingKnobs.BeforeWaitInRetryJobsWithExponentialBackoffMigration(jobspb.JobID(mutation.JobID))
+				}
+				if _, err := d.InternalExecutor.Exec(ctx, "migration-mutations-wait",
+					nil, "SHOW JOB WHEN COMPLETE $1", mutation.JobID); err != nil {
+					return err
+				}
+			}
+			continue
+		}
+
+		// Ignore the schema change if the table already has the required schema.
+		if ok, err := schemaExists(jt); err != nil {
+			return errors.Wrapf(err, "error while validating descriptors during"+
+				" operation %s", opTag)
+		} else if ok {
+			log.Infof(ctx, "skipping %s operation as the schema change already exists.", opTag)
+			// TODO(sajjad): To discuss: This seems not the best way to ensure that we
+			// are ignoring a mutation. We should have some other direct invariant,
+			// e.g., reading some counter from the table descriptor or in-memory state
+			// that counts the number of times the table has been successfully updated.
+			if d.TestingKnobs != nil && d.TestingKnobs.SkippedMutation != nil {
+				d.TestingKnobs.SkippedMutation()
+			}
+			return nil
+		}
+
+		// Modify the table.
+		log.Infof(ctx, "performing operation: %s", opTag)
+		if _, err := d.InternalExecutor.ExecEx(
+			ctx,
+			fmt.Sprintf("migration-alter-table-%d", tableID),
+			nil, /* txn */
+			sessiondata.InternalExecutorOverride{User: security.NodeUserName()},
+			schemaChangeQuery); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func readTableDescriptor(
+	ctx context.Context, d migration.TenantDeps, tableID descpb.ID,
+) (catalog.TableDescriptor, error) {
+	var jt catalog.TableDescriptor
+	if err := descs.Txn(ctx, d.Settings, d.LeaseManager, d.InternalExecutor, d.DB, func(
+		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
+	) (err error) {
+		jt, err = descriptors.GetImmutableTableByID(ctx, txn, tableID, tree.ObjectLookupFlags{
+			CommonLookupFlags: tree.CommonLookupFlags{
+				AvoidCached: true,
+				Required:    true,
+			},
+		})
+		return err
+	}); err != nil {
+		return nil, err
+	}
+	return jt, nil
+}
+
+// hasColumn return true if storedTable already has the given column. storedTable
+// descriptor must be read from the system as compared to reading from the
+// systemschema package.
+// This function returns an error if the column exists but doesn't match with the
+// table's descriptor defined in systemschema/system.go. The comparison is not strict
+// as several descriptor fields are ignored.
+func hasColumn(storedTable catalog.TableDescriptor, colName string) (bool, error) {
```

Comment:
can you call this something about the jobs table or pass the jobs table as a parameter?